### ISSUE TITLE
Fast Walsh–Hadamard transform for SciPy ?

### ISSUE THREAD

**Issue by ChamanAgrawal** [scipy/scipy#8590]
Currently there is no implementation of Fast Walsh–Hadamard transform in SciPy. Although it seems that FWHT is not as general as FFT but it is pretty ubiquitous ,it is there is other maths and science softwares like MATLAB etc. .I would like to contribute towards it. Following are the details about it.

## Fast Walsh–Hadamard transform
Hadamard transform is an example of a generalized class of Fourier transforms. It performs an orthogonal, symmetric, involutive, linear operation on 2^(m) numbers. It is equivalent to a multidimensional DFT.

### Time Complexity:
O(nlogn) with Fast Walsh-Hadamard transform (FWHT)

### Comparison with FFT:
 FWHT is very useful for reducing bandwidth storage requirements and spread-spectrum analysis. Compared to the FFT, the FWHT requires less storage space and is faster to calculate because it uses only real additions and subtractions, while the FFT requires complex values. The FWHT is able to represent signals with sharp discontinuities more accurately using fewer coefficients than the FFT.

### Some Usage examples:
The Hadamard transform is used in data encryption, as well as many signal processing and data compression algorithms, such as JPEG XR and MPEG-4 AVC. It is also a crucial part of Grover's algorithm and Shor's algorithm in quantum computing. The Hadamard transform is also applied in scientific methods such as NMR, mass spectroscopy and crystallography.

**Comment by ChamanAgrawal** [comment]
@ilayn Then should I start implementing this.

**Comment by ilayn** [comment]
That would be great indeed. But I would really recommend first checking out how scipy.signal library is structured. Development incentive and time are too valued hence better to streamline before you dedicate a significant effort. 

I would first start trying out a few fwht code lying around on GitHub and decide whether the speed is acceptable and if we need sequency and natural ordering etc. Some planning ahead would save us from a lot of discussion afterwards.

**Comment by rgommers** [comment]
I'd suggest discussing this on the mailing list first, that's where we decide if it's a good fit for SciPy (probably yes) and where it should go (my opinion would be `fftpack` rather than `signal`).

**Comment by rgommers** [comment]
https://github.com/stevengj/Hadamard.jl looks good, valuable as a reference and to generate test data.

**Comment by ChamanAgrawal** [comment]
Thanks @rgommers for the reference, btw should the implementation be for arbitrary dimensional data or it should be for 1-D,2-D and 3-D as they are the most common once.

**Comment by rgommers** [comment]
An n-D version would be nice. You can compare with how things are done for FFT - there's functions for 1-D (`fft` & co), 2-D (`fft2` & co), and n-D (`fftn` & co).

**Comment by ChamanAgrawal** [comment]
Can I implement it in cython or C ,I am not familiar with fortran to implement it like FFT is implemented.
Also I will be really helpful if I can be provided with few instructions regarding the structure that will help me to implement FWHT .

**Comment by ilayn** [comment]
Yes, you can use Cython or C. That would be not an issue. And you can start by mimicking a PR that uses a similar workflow, say, #8541, and then a few trial and error would take you there.

If you bring this intention in the mailing list there are some QC people occasionally visiting and might also help us to get the required complete feature list that is expected from such a function.

**Comment by ChamanAgrawal** [comment]
Hi I will be little late for the implementation but I wanted to discuss my idea before implementation -
1) User can give as input a multidimensional array along with the dimensions along which he wants the transformation.
2) There will be three ordering for transformation - natural , sequency and dyadic.
3) There will be three functions fht1,fht2,fhtn.
4) For n dimensions I have thought to implement in the same FFT is implemented i.e. by flattening the array.
5) Flattening will be done to create a 2D array with one dimension formed from those in which transformation is required and other from the remaining.
6) For flattening in fhtn I will swap axes to have all transforming dimensions at the start and so after flattening in the resultant 2D array I will have to apply fht2 along the 0th axis.Then I will reshape the 2D array to get the old shape and then deswap to get the answer.

In the 6th step I have a doubt that I want to ask with an example-
Suppose there is a 5D array and transformation is required in (1,3,4)axes then I can swap axes in this order(0-1 then 0-3 then 2-4) to get required order (1,3,4,0,2) then I can apply swap in the same order or reverse order to get the old ordering , but what to same order swap or reversed order swap.Will it affect my answer?

**Comment by ilayn** [comment]
By the way here is another MIT licensed C++ implementation, just in case : https://github.com/Sheljohn/WalshHadamard

**Comment by ChamanAgrawal** [comment]
Yeah thanks but It is for full transformation which in not a problem.Difficulty is in applying transformation along custom axes as there was in [https://github.com/stevengj/Hadamard.jl](url) suggested by @rgommers

**Comment by ChamanAgrawal** [comment]
Sorry done by mistake
Yeah thanks but It is for full transformation which in not a problem.Difficulty is in applying transformation along custom axes as there was in [https://github.com/stevengj/Hadamard.jl](url) suggested by @rgommers

**Comment by alphatangojuliett** [comment]
@ChamanAgrawal , any update on if an FWHT will make it to scipy soon? This would be quite useful.

**Comment by ChamanAgrawal** [comment]
Actually, I was not able to how to do the partial transformation in case n-dimensional input so I was stuck as I recall. I tried to search for it but was unable to find any helpful resources.

**Comment by alehresmann** [comment]
One can look to FFHT: https://github.com/FALCONN-LIB/FFHT for an optimised implementation of FHT which also provides a thin python wrapper for numpy arrays.

**Comment by ChamanAgrawal** [comment]
@alehresmann yeah, this seems to be a good reference but I am extremely busy this month with my GSoC so will be slow doing this.

**Comment by thomasahle** [comment]
Fast Hadamard transform is now in `sympy`: https://docs.sympy.org/latest/modules/discrete.html#fast-walsh-hadamard-transform

**Comment by nonohuff** [comment]
Does this implementation have orderings (natural , sequency and dyadic) as a parameter?
